File: stack/155_MinStack.go
package stack

import "math"

type MinStack struct {
	stack    []int
	minStack []int
}

func Constructor() MinStack {
	return MinStack{
		stack:    make([]int, 0),
		minStack: make([]int, 0),
	}
}

func (this *MinStack) Push(val int) {
	minVal := math.MaxInt32
	if len(this.minStack) > 0 {
		minVal = this.minStack[len(this.minStack)-1]
	}
	if val < minVal {
		minVal = val
	}

	this.minStack = append(this.minStack, minVal)
	this.stack = append(this.stack, val)
}

func (this *MinStack) Pop() {
	this.minStack = this.minStack[:len(this.minStack)-1]
	this.stack = this.stack[:len(this.stack)-1]
}

func (this *MinStack) Top() int {
	return this.stack[len(this.stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.minStack[len(this.minStack)-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: stack/155_MinStack.py
class MinStack:
    def __init__(self):
        # holds all values on stack like regular stack would
        self.stack = []
        # holds minimum value on a `stack` up to this point in stack
        # in other words each node in stack has a minimum element assigned to it
        self.min_stack = []

    def push(self, val: int) -> None:
        if self.min_stack:
            min_val = self.min_stack[-1]
        else:
            min_val = float("inf")

        min_val = min(min_val, val)

        self.stack.append(val)
        self.min_stack.append(min_val)

    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
